drivers/adc/adc_stm32.c

```diff
@@ -1347,6 +1348,117 @@ static int adc_stm32_channel_setup(const struct device *dev,
 	return 0;
 }
 
+#if defined(CONFIG_SOC_SERIES_STM32C0X) || defined(CONFIG_SOC_SERIES_STM32G0X) ||                  \
+	defined(CONFIG_SOC_SERIES_STM32L0X) ||                                                     \
+	(defined(CONFIG_SOC_SERIES_STM32WBX) && defined(ADC_SUPPORT_2_5_MSPS)) ||                  \
+	defined(CONFIG_SOC_SERIES_STM32WLX) || defined(CONFIG_SOC_SERIES_STM32H7X)
+static bool adc_stm32_is_clk_sync(const struct adc_stm32_cfg *config)
+{
+	if (config->clk_prescaler == LL_ADC_CLOCK_SYNC_PCLK_DIV1 ||
+	    config->clk_prescaler == LL_ADC_CLOCK_SYNC_PCLK_DIV2 ||
+	    config->clk_prescaler == LL_ADC_CLOCK_SYNC_PCLK_DIV4) {
+		return true;
+	}
+
+	return false;
+}
+#endif
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+static inline int adc_stm32_get_input_freq_prescaler(void)
+{
+	int presc = 2;
+
+	/* For revision Y we have no prescaler of 2 */
+	if (LL_DBGMCU_GetRevisionID() <= 0x1003) {
+		presc = 1;
+	}
+
+	return presc;
+}
+
+static int adc_stm32_get_clock_prescaler(const struct adc_stm32_cfg *config)
+{
+	switch (config->clk_prescaler) {
+	case LL_ADC_CLOCK_SYNC_PCLK_DIV1:
+	case LL_ADC_CLOCK_ASYNC_DIV1:
+		return 1;
+	case LL_ADC_CLOCK_SYNC_PCLK_DIV2:
+	case LL_ADC_CLOCK_ASYNC_DIV2:
+		return 2;
+	case LL_ADC_CLOCK_SYNC_PCLK_DIV4:
+	case LL_ADC_CLOCK_ASYNC_DIV4:
+		return 4;
+	case LL_ADC_CLOCK_ASYNC_DIV6:
+		return 6;
+	case LL_ADC_CLOCK_ASYNC_DIV8:
+		return 8;
+	case LL_ADC_CLOCK_ASYNC_DIV10:
+		return 10;
+	case LL_ADC_CLOCK_ASYNC_DIV12:
+		return 12;
+	case LL_ADC_CLOCK_ASYNC_DIV16:
+		return 16;
+	case LL_ADC_CLOCK_ASYNC_DIV32:
+		return 32;
+	case LL_ADC_CLOCK_ASYNC_DIV64:
+		return 64;
+	case LL_ADC_CLOCK_ASYNC_DIV128:
+		return 128;
+	case LL_ADC_CLOCK_ASYNC_DIV256:
+		return 256;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int adc_stm32h7_setup_boost(const struct adc_stm32_cfg *config, ADC_TypeDef *adc,
+				   const struct device *clk)
+{
+	clock_control_subsys_t clk_src;
+	uint32_t input_freq;
+	uint32_t boost;
+	int presc;
+
+	/* Get the input frequency */
+	clk_src = (clock_control_subsys_t)(adc_stm32_is_clk_sync(config) ? &config->pclken[0]
+									 : &config->pclken[1]);
```

Comment:
Another thing I just noticed, config->pclken[1] does not necessarily exist. It is currently possible to be in ASYNC but not defining a secondary clock source. In that case, PLL2_P will be used (default one). But I agree it is not a good practice.

Maybe you could add a check (not limited to H7) to generate a compile-time error if ASYNC is defined without an explicit secondary clock.

But you still have to update this line for cases in SYNC that may not have the secondary clock.

Reply:
@gautierg-st I did not understand the last line.. Why does it matter if pclken[1] is not set when running in SYNC mode ? AFAICS we only need to check pclken[1] when ASYNC mode is used ?

Reply:
Well if you add global check, then yeah, your current code should be ok.

But thinking about it, I believe a simple check like the following could suffice:
```
clk_src = &config->pclken[0];
if (IS_ENABLED(STM32_ADC_DOMAIN_CLOCK_SUPPORT) && (config->pclk_len > 1) && !adc_stm32_is_clk_sync(config)) {
	clk_src = &config->pclken[1];
}
```

Reply:
> Well if you add global check, then yeah, your current code should be ok.
> 
> But thinking about it, I believe a simple check like the following could suffice:
> 
> ```
> clk_src = &config->pclken[0];
> if (IS_ENABLED(STM32_ADC_DOMAIN_CLOCK_SUPPORT) && (config->pclk_len > 1) && !adc_stm32_is_clk_sync(config)) {
> 	clk_src = &config->pclken[1];
> }
> ```

I have chosen for a compile error check as that code is running on H7 only anyway. Please check the current version ;)